app/src/main/java/samyups/example/trainingss/data/MainRepository.kt

```diff
@@ -1,31 +1,15 @@
 package samyups.example.trainingss.data
 
-import androidx.lifecycle.MutableLiveData
+import androidx.annotation.WorkerThread
+import kotlinx.coroutines.flow.Flow
 
-class MainRepository private constructor () {
+class MainRepository(private val mainDao: MainDao) {
 
-    private val dogsList = mutableListOf<Dog>()
-    private val dogs = MutableLiveData<List<Dog>>()
+    val dogsList : Flow<List<Dog>> = mainDao.getDogList()
 
-    init {
-        dogs.value = dogsList
+    @Suppress("RedundantSuspendModifier")
+    @WorkerThread
+    suspend fun addDog(dog: Dog) {
+        mainDao.addDog(dog)
     }
-
-    fun getDogsList() = dogs
-
-    fun addDog(dog: Dog) {
-
-        dogsList.add(dog)
-        dogs.value = dogsList
-    }
-
-
-    companion object {
-        @Volatile private var instance : MainRepository? = null
-
-        fun getInstance() =
-            instance ?: synchronized(this) {
-                instance ?: MainRepository().also { instance = it }
-            }
-    }
-}
\ No newline at end of file
+}  // How come I don't need a singleton for this repository?
```

Comment:
A Singleton just means that you only have one instance of a class that is used by the entire application. The reason for this is usually because the object contains some kind of data that the entire application needs. This means that the object contains `state` that is needed by the entire application. `state` is just data, like the current state of something. 

Think of it this way, if Sammy is a Singleton, that means there is only one copy of Sammy. And if Sammy is sad, his current state is sad. If Sammy is happy, his current state is happy. Anyone that calls Sammy on his phone and asks how he is doing, he will give the same answer. If he's happy he will tell anyone who asks that he is happy. If he is sad, he will tell anyone who asks he is sad. Because that is his current state, and there is only one Sammy and anyone who calls him up talks to the same Sammy. Now, if there are multiple clones of Sammy running around, and one is happy and one is sad, depending on which Sammy you talk to you will get a different answer when you ask him how he is doing. In this case, it's important that Sammy is a Singleton because we care about Sammy's feelings and we want to know how he is truly feeling. Now if we had a robot named Manny, we don't give a shit how the robot feels. In fact, robots don't have feelings. So you can't even ask Manny how he is feeling. You can just tell him to do stuff and he will do it. So if I call Manny and say bring me coffee, and someone else calls Manny and says bring me tea, it doesn't matter if we call the same Manny. Or if there are two separate copies of Manny doing the work. He contains no state. 

Previously you were keeping `List<Dog>>` in memory within this `Repository` (i.e. keeping the list of dogs stored in a variable that lived in this class). Since you were keeping data in this object, it means this object contained `state` that your application needed to work. So it should be a Singleton since the entire application needs to reference the same state. If you weren't using a Singleton, and you had multiple copies of `Repository` floating around, how do you keep track of which copy gets updated and which doesn't? Do you update all copies? It starts to get pretty messy. So by having a Singleton, you guarantee that the data lives in one place and anyone accessing it is accessing the same data. 

Now, with this database, you aren't keeping `state` in this class. You have methods that read and write data from a database, but this class in particular is `stateless`. No data of any kind is stored in this class that is needed for your application to function. Instead it is kind of like a pass-through piece of code. It gets data from one place (your database) and hands it off to somewhere else (your `viewmodel`), but this class itself contains no data. So because of that, you don't need one single instance of this class to exist. Each part of your code could be using it's own instance of this class, and it wouldn't matter because this class only contains functionality, not data. And that functionality is the same no matter how many instances you create and who uses those instances. 

Does that make sense? This is actually a very important concept when you start to get into more advanced programming techniques, so you are thinking the correct things, and asking the right questions, so good job! We generally refer to this as the Scope of an object and you will have to get familiar with it when you start writing more complicated code, but even then not for a while. Dagger2 is the library we use to help us create the right scopes and stuff. For now I think it is not super important for you to know scopes and Dagger2 inside out, it is considered an intermediate/advanced concept and I would be surprised if anyone expected you to know about it at this stage. But you're asking the right questions, so keep this knowledge in your back pocket, it will be useful.